staging/src/k8s.io/dynamic-resource-allocation/structured/namedresources/cel/compile.go

```diff
@@ -0,0 +1,228 @@
+	}
+}
+
+var valueTypes = map[string]struct {
+	celType *cel.Type
+	// get returns nil if the attribute doesn't have the type, otherwise
+	// the value of that type.
+	get func(attr resourceapi.NamedResourcesAttribute) (any, error)
+}{
+	"quantity": {apiservercel.QuantityType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.QuantityValue == nil {
+			return nil, nil
+		}
+		return apiservercel.Quantity{Quantity: attr.QuantityValue}, nil
+	}},
+	"bool": {cel.BoolType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.BoolValue == nil {
+			return nil, nil
+		}
+		return *attr.BoolValue, nil
+	}},
+	"int": {cel.IntType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.IntValue == nil {
+			return nil, nil
+		}
+		return *attr.IntValue, nil
+	}},
+	"intslice": {types.NewListType(cel.IntType), func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.IntSliceValue == nil {
+			return nil, nil
+		}
+		return attr.IntSliceValue.Ints, nil
+	}},
+	"string": {cel.StringType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.StringValue == nil {
+			return nil, nil
+		}
+		return *attr.StringValue, nil
+	}},
+	"stringslice": {types.NewListType(cel.StringType), func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.StringSliceValue == nil {
+			return nil, nil
+		}
+		return attr.StringSliceValue.Strings, nil
+	}},
+	"version": {SemVerType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.VersionValue == nil {
+			return nil, nil
+		}
+		v, err := semver.Parse(*attr.VersionValue)
+		if err != nil {
+			return nil, fmt.Errorf("parse semantic version: %v", err)
+		}
+
+		return SemVer{Version: v}, nil
+	}},
+}
+
+var boolType = reflect.TypeOf(true)
+
+func (c CompilationResult) Evaluate(ctx context.Context, attributes []resourceapi.NamedResourcesAttribute) (bool, error) {
+	variables := make(map[string]any, len(valueTypes))
+	for name, valueType := range valueTypes {
+		m, err := buildValueMapper(c.Environment.CELTypeAdapter(), attributes, valueType.get)
+		if err != nil {
+			return false, fmt.Errorf("extract attributes with type %s: %v", name, err)
+		}
+		variables[attributesVarPrefix+name] = m
+	}
+	result, _, err := c.Program.ContextEval(ctx, variables)
+	if err != nil {
+		return false, err
+	}
+	resultAny, err := result.ConvertToNative(boolType)
+	if err != nil {
+		return false, fmt.Errorf("CEL result of type %s could not be converted to bool: %w", result.Type().TypeName(), err)
+	}
+	resultBool, ok := resultAny.(bool)
+	if !ok {
+		return false, fmt.Errorf("CEL native result value should have been a bool, got instead: %T", resultAny)
+	}
+	return resultBool, nil
+}
+
+func mustBuildEnv() *environment.EnvSet {
+	envset := environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion())
+	versioned := []environment.VersionedOptions{
+		{
+			IntroducedVersion: version.MajorMinor(1, 30),
```

Comment:
in other places where we've added a CEL library and a consuming API in the same release, we've done this:

```
			// Feature epoch was actually 1.23, but we artificially set it to 1.0 because these
			// options should always be present.
			IntroducedVersion: version.MajorMinor(1, 0),
```

since it's impossible for namedresources to use CEL prior to 1.30, there's no need to guard the semver cel functions added in 1.30

Reply:
Agreed. I'd like 1.0 to be used to represent "introduced at the same time the feature using it was introduced" here as well.

Reply:
I like my approach with always using `StoredExpression` when selecting the environment better, see https://github.com/kubernetes/kubernetes/pull/123516/files#r1516514152

Both works and I can pivot to whatever you prefer.

Reply:
> I like my approach with always using StoredExpression when selecting the environment better

That will do the wrong thing in the future if you add more CEL functions later

Reply:
But it won't matter while we are in alpha, will it? We always enable everything, either way.

Then once we want proper versioning, that hack in validation.go must go away and all CEL env features must be properly versionined.

Reply:
> That will do the wrong thing in the future if you add more CEL functions later

+1.  Using 1.0 allows use of NewExpressions/StoredExpressions [exactly as defined](https://github.com/kubernetes/kubernetes/blob/246e678acc97d68261ce3e9ae858332f01b44058/staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go#L39-L60) preventing problems with future usage library changes or usages of this environment

Reply:
Done and pushed.